### DISCUSSION TITLE
Augmenting Built-in ON Clauses, with no available relations

### DISCUSSION THREAD

**Discussion by bigjools** [sqlalchemy/sqlalchemy#6831] [Usage Questions] [answered]
I'm following the documents at https://docs.sqlalchemy.org/en/14/orm/queryguide.html#augmenting-built-in-on-clauses
which details the new 1.4 ` PropComparator.and_()`.

If you don't use any `and_` and there are no matching joins on that property, the whole row is still returned. If I add an `and_` that ends up matching nothing, then the whole row is redacted from the results like you'd expect with a normal JOIN.

I was hoping that this new feature would *only* filter the related rows. As it stands it's not much use over a normal JOIN, where I would use an `outerjoin` in this case where the RHS doesn't match.

Is this how it's expected to work or am I doing it wrong?

Example:

```python
from sqlalchemy import event, create_engine, Column, Integer, ForeignKey
from sqlalchemy.orm import declarative_base, sessionmaker, relationship
from sqlalchemy.sql.expression import select

Base = declarative_base()
class Foo(Base):
    __tablename__ = "foo"
    id = Column(Integer, primary_key=True)
    foo = Column(Integer)
    bars = relationship('Bar')

class Bar(Base):
    __tablename__ = "bar"
    id = Column(Integer, primary_key=True)
    bar = Column(Integer)
    foo_id = Column(Integer, ForeignKey('foo.id'))
    foo = relationship(Foo, back_populates='bars')


engine = create_engine("sqlite://", future=True)
Session = sessionmaker()
Base.metadata.create_all(bind=engine)

connection = engine.connect()
txn = connection.begin()
session = Session(bind=connection)

session.add(foo := Foo(foo=1))
session.add(foo := Foo(foo=2))
session.add(Bar(bar=1, foo=foo))
session.add(Bar(bar=2, foo=foo))
session.commit()

q = select(Foo).join(Foo.bars.and_(Bar.bar == 3))
print("select(Foo).join(Foo.bars.and_(Bar.bar == 3))")
print(session.execute(q).all())

q = select(Foo).where(Foo.foo == 2)
print("select(Foo).where(Foo.foo == 2)")
print(session.execute(q).all())

q = select(Foo).join(Foo.bars.and_(Bar.bar.in_([1, 2])))
results = session.execute(q).all()
print("select(Foo).join(Foo.bars.and_(Bar.bar.in_([1, 2])))")                                                                                                                                                                  
print(results)
for row in results:
    print(row[0].bars)
```

**Comment by lesnik512** [comment, marked as answer]
I prefer to use [selectinload](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#sqlalchemy.orm.selectinload) with [with_loader_criteria](https://docs.sqlalchemy.org/en/14/orm/query.html#sqlalchemy.orm.with_loader_criteria) for such task

```python
from sqlalchemy import event, create_engine, Column, Integer, ForeignKey
from sqlalchemy.orm import declarative_base, sessionmaker, relationship, selectinload, with_loader_criteria
from sqlalchemy.sql.expression import select


Base = declarative_base()
class Foo(Base):
    __tablename__ = "foo"
    id = Column(Integer, primary_key=True)
    foo = Column(Integer)
    bars = relationship('Bar')

class Bar(Base):
    __tablename__ = "bar"
    id = Column(Integer, primary_key=True)
    bar = Column(Integer)
    foo_id = Column(Integer, ForeignKey('foo.id'))
    foo = relationship(Foo, back_populates='bars')


engine = create_engine("sqlite://", future=True)
Session = sessionmaker()
Base.metadata.create_all(bind=engine)

connection = engine.connect()
txn = connection.begin()
session = Session(bind=connection)

session.add(foo1 := Foo(foo=1))
session.add(foo2 := Foo(foo=2))
session.add(Bar(bar=1, foo=foo1))
session.add(Bar(bar=2, foo=foo2))
session.commit()

q = select(Foo).options(selectinload(Foo.bars), with_loader_criteria(Bar, Bar.bar == 1))
result = session.execute(q).scalars().all()
print(result[0].bars, result[1].bars)  # [<__main__.Bar object at 0x7f8085679730>] []
```

**Reply by bigjools** [reply]
I figured `selectinload` would feature somehow, thanks for the pointer!

**Reply by zzzeek** [reply]
didnt follow this too closely but if you are looking to affect the relationship load, that is `foo.bars`, then you need to be using a loader option like selectinload.   but I would note the above example can be written using `selectinload(Foo.bars.and_(Bar.bar == 1))` directly, no need for the extra with_loader_criteria part:

    # equivalent to q = select(Foo).options(selectinload(Foo.bars), with_loader_criteria(Bar, Bar.bar == 1)) in this case
 
    q = select(Foo).options(selectinload(Foo.bars.and_(Bar.bar == 1)))

**Reply by bigjools** [reply]
Oh that's EXACTLY the kind of response I was looking for, thanks Mike. It might be worth updating the docs to clarify this as it was not obvious to me how `and_` would exactly affect the results until I experimented.

**Reply by bigjools** [reply]
Does applying these options work with the old Query API as well? I'm unable to get it to work there so far (I'm in the middle of converting all my application to use Selects, but there's a lot).

**Reply by zzzeek** [reply]
all of this stuff works exactly the same way with the legacy Query API.   the legacy Query object just creates the new style select() anyway.

**Reply by bigjools** [reply]
Ok I will try to make a small non-working example as I'm seeing it completely ignore the option. The outer query is selecting multiple models in the return tuple so I wonder if that's affecting things.

**Reply by zzzeek** [reply]
keep in mind query options only take effect for the outermost query, if you call subquery() on something, those options inside there are ignored.

**Reply by bigjools** [reply]
This is a snippet of the exact code that doesn't work. I hacked in the `[999]` as a test but normally that's a `select`.
```python
        sources_option = selectinload(
            models.Dependency.sources.and_(
                models.DependencySource.id.in_([999])
            )
        )
        q = q.options(sources_option)
```
The outermost query is doing a `session.query(models.Dependency, func, func, func)` where those are three group functions. So it should just be affecting the `sources` property on a model in the outermost query.

**Reply by zzzeek** [reply]
here's your script, try it out and see how yours is different (did you expire the objects and/or use populate_existing() ?)

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import func
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import selectinload
from sqlalchemy.orm import Session

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)
    bs = relationship("B")


class B(Base):
    __tablename__ = "b"
    id = Column(Integer, primary_key=True)
    a_id = Column(ForeignKey("a.id"))
    data = Column(String)


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(A(bs=[B(data="b1"), B(data="b2"), B(data="b3")]))
s.commit()

for row in s.query(A, func.count("*")).group_by(A.id):
    assert len(row[0].bs) == 3
    print(row[0].bs)

# the important bit if you are limiting eagerly loaded collections!
s.expire_all()

sources_option = selectinload(A.bs.and_(B.data.in_(["b1", "b2"])))

for row in s.query(A, func.count("*")).group_by(A.id).options(sources_option):
    assert len(row[0].bs) == 2
    print(row[0].bs)
```

**Reply by bigjools** [reply]
I'm querying on an existing database so should not need to expire, but I added it and it makes no difference. I added an echo to the engine to get the SQL logged, and it's just running the standard WHERE IN for the relation. Scratching my head a bit here, so I'll try to boil it down to a minimal example. Either I will see where my PEBKAC is or I'll find a bug :)

**Reply by bigjools** [reply]
Can confirm PEBKAC on my part - or at least the idiosyncrasies of a rather large web app. Thanks for the help!

**Reply by bigjools** [reply]
I hit another wall, and I am not sure this feature is going to work, unless there's some dark magic I can intercept. My selectin clause needs to refer to columns in the main query, but since it's run separately, it would need to transfer values from that main query similar to how selectin is working with PKs by default. Is this possible? I tried a joinedload but that failed miserably since the table I need is buried in a subselect (arising because of an `exists` check).

**Reply by zzzeek** [reply]
it would be easier here then if you ran the query that you intend for `selectinload` explicitly instead; load the objects, then knit together the collections manually.   when you have a collection of "B" objects to go inside a particular "A.bs" collection, use [set_committed_value](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=set_committed_value#sqlalchemy.orm.attributes.set_committed_value) to set the list attribute.

we used to have an example of this and im not really sure why it vanished will see if i can find it.

**Reply by zzzeek** [reply]
updated one here: https://github.com/sqlalchemy/sqlalchemy/wiki/DisjointEagerLoading

**Reply by bigjools** [reply]
Oh thanks Mike, I had not seen `set_committed_value` before! That's actually quite handy and your example is _almost_ doing what my code was doing pre-1.4 anyway which was using `string_agg` functions to pull out related collection PKs from an outer join, and mass load them separately.

Once again, many thanks.
